Add tutorial for GPUs on EKS Auto Mode (#914)
awsdocs/amazon-eks-user-guide: latest/ug/automode/auto-accelerated.adoc

Apply the NodePool and workload to your cluster.

[source,bash]
----
kubectl apply -f nodepool-gpu.yaml
kubectl apply -f pod.yaml
----

You should see the following output:

[source,bash]
----
nodepool.karpenter.sh/gpu configured created
pod/nvidia-smi created
----

Wait a few seconds, and check the nodes in your cluster. You should now see a new node provisioned in your Amazon EKS Auto Mode cluster:

[source,bash]
----
> kubectl get nodes

NAME        TYPE          CAPACITY    ZONE         NODE                  READY   AGE
gpu-dnknr   g6e.2xlarge   on-demand   us-west-2b   i-02315c7d7643cdee6   True    76s
----

== Step 2: Validate

You can see Amazon EKS Auto Mode launched a `g6e.2xlarge` rather than an `g6.2xlarge` as the workload required an instance with l40s `GPU`, according to the following Kubernetes scheduling constraints:

[source,yaml]
----
...
  nodeSelector:
    eks.amazonaws.com/instance-gpu-name: l40s
...
    requests:
        memory: "30Gi"
        cpu: "3500m"
        nvidia.com/gpu: 1
      limits:
        memory: "30Gi"
        nvidia.com/gpu: 1
----

Now, look at the containers logs, by running the following command:

----

> `>`` kubectl logs nvidia``-``smi``

Reviewer: Looks like formatting is broken? This should be kubectl logs nvidia-smi.

Author: good catch

[.topic]
[#auto-accelerated]
= Deploy an accelerated workload
:info_titleabbrev: Deploy accelerated workload

include::../attributes.txt[]

> This tutorial demonstrates how Amazon EKS Auto Mode simplifies launching GPU-based workloads. Amazon EKS Auto Mode streamlines operations beyond the cluster itself by automating key infrastructure components providing compute, networking, load balancing, storage, and Identity Access and Management capabilities out of the box. 

Reviewer 2: ```suggestion
This tutorial demonstrates how Amazon EKS Auto Mode simplifies launching accelerated workloads. Amazon EKS Auto Mode streamlines operations beyond the cluster itself by automating key infrastructure components providing compute, networking, load balancing, storage, and Identity Access and Management capabilities out of the box. 
```

> Amazon EKS Auto Mode includes the drivers required for certain instance types, such as NVIDIA and {aws} Neuron drivers. You do not have to install or update these drivers. 

Reviewer 2: ```suggestion
Amazon EKS Auto Mode includes the drivers and device plugins required for certain instance types, such as NVIDIA and {aws} Neuron drivers. You do not have to install or update these components. 
```

EKS Auto Mode manages the drivers for: 

* link:ai/machine-learning/trainium/[{aws} Trainium, type="marketing"]
* link:ai/machine-learning/inferentia/[{aws} Inferentia, type="marketing"]
* NVIDIA GPU link:ec2/latest/instancetypes/ac.html[Amazon EC2 accelerated instances,type="documentation"]

> ** Including the https://github.com/NVIDIA/k8s-device-plugin[NVIDIA device plugin for Kubernetes]. EKS Auto Mode includes this plugin, and is not visible as a daemon set within your cluster. 

Reviewer 2: we include device plugins for Neuron (Trainium and Inferentia) as well as EFA in addition to the NVIDIA device plugin

Amazon EKS Auto Mode also includes support for link:hpc/efa/[Elastic Fabric Adapter,type="marketing"].

> //GDC: What does the EFA thing mean?

Reviewer 2: its a high speed, high throughput, low latency network used within HPC and machine learning applications where the workload spans multiple nodes (too large to fit on a single node). Its Amazon's version of Infiniband

> Amazon EKS Auto Mode eliminates the toil of GPU driver management. 

Reviewer 2: ```suggestion
Amazon EKS Auto Mode eliminates the toil of accelerator driver and device plugin management. 
```

You can also benefit from cost savings by scaling the cluster to zero. You can configure EKS Auto Mode to terminate instances when no workloads are running. This is useful for batch based inference workloads.

> The following provides an example of how to launch GPU-based workloads with Amazon EKS Auto Mode.

Reviewer 2: ```suggestion
The following provides an example of how to launch accelerated workloads with Amazon EKS Auto Mode.
```

== Prerequisites

* A Kubernetes cluster with Amazon EKS Auto Mode configured.
* A `default` EKS Node class as created when the `general-purpose` or `system` Managed Node Pools are enabled.

== Step 1: Deploy a GPU workload

> In this example, you will create a NodePool for NVIDIA inference based workloads that requires 45GB GPU memory. With EKS Auto Mode, you use Kubernetes scheduling constraints to define your instance requirements. 

Reviewer 2: the phrasing was a bit long, and theres nothing specific to inference
```suggestion
In this example, you will create a NodePool for NVIDIA based workloads that requires 45GB GPU memory. With EKS Auto Mode, you use Kubernetes scheduling constraints to define your instance requirements. 
```

Reviewer: Suggested:

"In this example, you will create a NodePool for NVIDIA based workloads that requires `l40s` GPU. With EKS Auto Mode, you use Kubernetes scheduling constraints to define your instance requirements."

To deploy the Amazon EKS Auto Mode `NodePool` and the sample `workload`, review the following NodePool and Pod definition and save as `nodepool-gpu.yaml` and `pod.yaml`:

*nodepool-gpu.yaml*

[source,yaml]
----
apiVersion: karpenter.sh/v1
kind: NodePool
metadata:
  name: gpu
spec:
  disruption:
    budgets:
    - nodes: 10%
    consolidateAfter: 1h
    consolidationPolicy: WhenEmpty
  template:
    metadata: {}
    spec:
      nodeClassRef:
        group: eks.amazonaws.com
        kind: NodeClass
        name: default
      requirements:
        - key: "karpenter.sh/capacity-type"
          operator: In
          values: ["on-demand"]
        - key: "kubernetes.io/arch"
          operator: In
          values: ["amd64"]
        - key: "eks.amazonaws.com/instance-family"
          operator: In
          values:
          - g6e
          - g6
      terminationGracePeriod: 24h0m0s
----

*pod.yaml*

[source,yaml]
----
apiVersion: v1
kind: Pod
metadata:
  name: nvidia-smi
spec:
  nodeSelector:

>     eks.amazonaws.com/instance-gpu-memory: "45776"

Reviewer 2: this will only match on g6e above, and given that GPU memory and arch of one GPU architecture is not equivalent to another, its more representative to use the card name itself (as in - "I want this pod to run on L40S GPUs") 

```suggestion
    eks.amazonaws.com/instance-gpu-name: l40s
```

    eks.amazonaws.com/compute-type: auto
  restartPolicy: OnFailure
  containers:
  - name: nvidia-smi

>     image: nvidia/cuda:12.8.0-cudnn-devel-ubuntu20.04

Reviewer 2: might be better to show `nvcr.io/nvidia/pytorch:25.02-py3` which has PyTorch instead of just a development image. This is a rather large image and will take time to pull. If we are just showing `nvidia-smi` then you can actually use something much smaller like `public.ecr.aws/amazonlinux/amazonlinux:2023-minimal`

Reviewer: +1, propose update to `public.ecr.aws/amazonlinux/amazonlinux:2023-minimal`

    args:
    - "nvidia-smi"
    resources:
      requests:
        memory: "30Gi"
        cpu: "3500m"
        nvidia.com/gpu: 1
      limits:
        memory: "30Gi"
        nvidia.com/gpu: 1
----

Note the `eks.amazonaws.com/compute-type: auto` selector requires the workload be deployed on an Amazon EKS Auto Mode node.

Apply the NodePool and workload to your cluster.

[source,bash]
----
kubectl apply -f nodepool-gpu.yaml
kubectl apply -f pod.yaml
----

You should see the following output:

[source,bash]
----
nodepool.karpenter.sh/gpu configured created
pod/nvidia-smi created
----

> Wait about one minute, and check the nodes in your cluster. You should now see a new node provisioned in your Amazon EKS Auto Mode cluster:

Reviewer 2: if we use the smaller image above, this will happen much faster, a couple of seconds

[source,bash]
----
> kubectl get nodes

NAME        TYPE          CAPACITY    ZONE         NODE                  READY   AGE
gpu-dnknr   g6e.2xlarge   on-demand   us-west-2b   i-02315c7d7643cdee6   True    76s
----

== Step 2: Validate

> You can see Amazon EKS Auto Mode launched a `g6e.2xlarge` rather than an `g6.2xlarge` as the workload required an instance with at least 45GB of GPU memory, according to the following Kubernetes scheduling constraints:

Reviewer: Suggest change:

"You can see Amazon EKS Auto Mode launched a `g6e.2xlarge` rather than an `g6.2xlarge` as the workload required an instance with `l40s` GPU, according to the following Kubernetes scheduling constraints:"

[source,yaml]
----
...
  nodeSelector:

>     eks.amazonaws.com/instance-gpu-memory: "45776"

Reviewer: Suggested to replace with: `eks.amazonaws.com/instance-gpu-name: l40s` as per above.

...
    requests:
        memory: "30Gi"
        cpu: "3500m"
        nvidia.com/gpu: 1
      limits:
        memory: "30Gi"
        nvidia.com/gpu: 1
----

Now, look at the containers logs, by running the following command:

----
`>`` kubectl logs nvidia``-``smi``

...
+---------------------------------------------------------------------------------------+

> | NVIDIA-SMI 535.230.02             Driver Version: 535.230.02   CUDA Version: 12.2     |

Reviewer 2: this seems quite dated - which AMI was used?

Author: is this dated beyond just the version numbers in the output?

if it's minimal differences, i'd like to get it out as is

Reviewer 2: sorry - more of a curiosity, not a blocker at all 😅

+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
...`
----

You can see that the container has detected it's running on an instance with an `NVIDIA` GPU and that you've not had to install any device drivers, as this is managed by Amazon EKS Auto Mode.

== Step 3: Clean-up

To remove all objects created, use `kubectl` to delete the sample deployment and NodePool so the node is terminated:

----
kubectl delete -f nodepool-gpu.yaml
kubectl delete -f pod.yaml
----

'''

> Files:

Reviewer: Lines 193 - 262 can be removed as covered earlier in the content.

----
kubectl delete -f .
----

== Example NodePools Reference

=== Create an NVIDIA NodePool

The following NodePool defines:

* Only launch instances of `g6e` and `g6` family
* Consolidate nodes when empty for 1 hour
** The 1 hour value for `consolodateAfter` supports spiky workloads and reduce node churn. You can tune `consolidateAfter` based on your workload requirements.

*Example NodePool with GPU instance family and consolidation*

[source,yaml]
----
apiVersion: karpenter.sh/v1
kind: NodePool
metadata:
  name: gpu
spec:
  disruption:
    budgets:
    - nodes: 10%
    consolidateAfter: 1h
    consolidationPolicy: WhenEmpty
  template:
    metadata: {}
    spec:
      nodeClassRef:
        group: eks.amazonaws.com
        kind: NodeClass
        name: default
      requirements:
        - key: "karpenter.sh/capacity-type"
          operator: In
          values: ["on-demand"]
        - key: "kubernetes.io/arch"
          operator: In
          values: ["amd64"]
        - key: "eks.amazonaws.com/instance-family"
          operator: In
          values:
          - g6e
          - g6
      terminationGracePeriod: 24h0m0s
----

> In addition, GPU workloads might be sensitive to specific GPU memory requirements, you might set a `eks.amazonaws.com/instance-gpu-memory` node selector on workloads so they schedule on nodes that have the required GPU memory resources. For other well-known labels which influence scheduling review, see <<auto-supported-labels>>.

Reviewer 2: see comment above - customers typically choose based off the GPU card type due to the NVIDIA compute capabilities of that card. this can be done by either specifying the instance family types or using something like `karpenter.k8s.aws/instance-gpu-name`

Reviewer: Suggestion:
"In addition, GPU workloads might be sensitive to specific GPU memory requirements, you might set a `eks.amazonaws.com/instance-gpu-memory` node selector along side `eks.amazonaws.com/instance-gpu-name` or `eks.amazonaws.com/instance-family` selector so they schedule on nodes that have the required GPU and resources. For other well-known labels which influence scheduling review, see <<auto-supported-labels>>"

Reviewer: Suggestion:

"Note: opposed to setting the `eks.amazonaws.com/instance-gpu-name` you might use `eks.amazonaws.com/instance-family` to specify the instance family. For other well-known labels which influence scheduling review, see ..."

> If you have specific storage requirements you can tune the nodes ephemeral storage `iops`, `size` and `throughput` by creating your own xref:create-node-class[NodeClass] to reference in the NodePool. Learn more about the xref:create-node-class[configurable NodeClass options].

Reviewer 2: I'm less familiar with this on Auto mode - I thought we were automatically setting `instanceStorePolicy: RAID0` which means that users cannot specify things like iops, size, etc. You just get the instance store volumes (found on our accelerated instance types) mounted in a RAID0 array - and thats it

Author: i'm gonna ask internally for a reviewer on this

*Example storage configuration for NodeClass*

[source,yaml]
----
apiVersion: eks.amazonaws.com/v1
kind: NodeClass
metadata:
  name: gpu
spec:
  ephemeralStorage:
    iops: 3000
    size: 80Gi
    throughput: 125
----

=== Define an {aws} Trainium and {aws} Inferentia NodePool

> The following NodePool has an `eks.amazonaws.com/instance-family` set that says, only launch instances of Inferentia and Trainium family:

Reviewer: Change `eks.amazonaws.com/instance-family` to `eks.amazonaws.com/instance-category`

----

>         - key: "eks.amazonaws.com/instance-family"

Reviewer 2: ```suggestion
        - key: "eks.amazonaws.com/instance-category"
```

which means below you can use
```yaml
          values:
            - inf
            - trn
```
